File: typescript/tests/affect_vad.test.ts
// affect_vad.test.ts
//
// Cross-language AffectVad derivation verification.
// Math must be byte-identical to the C# / Rust / Kotlin / Python ports.
//
// Derivation:
//   Valence   = (engagement + rapport + (1 - uncertainty)) / 3
//   Arousal   = (energy * 2 + curiosity + uncertainty) / 4
//   Dominance = (engagement + (1 - uncertainty)) / 2
// Outputs are clamped to [0.0, 1.0]. Compare with epsilon 1e-5.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { AffectState, affectStateToVad, type AffectVad } from '../src/memory';

// ---------------------------------------------------------------------------
// Vector types
// ---------------------------------------------------------------------------

interface AffectInput {
  curiosity:   number;
  engagement:  number;
  uncertainty: number;
  rapport:     number;
  energy:      number;
}

interface VadExpected {
  valence:   number;
  arousal:   number;
  dominance: number;
}

interface AffectVadVector {
  id:          string;
  description: string;
  input:       AffectInput;
  expected:    VadExpected;
}

// ---------------------------------------------------------------------------
// Vectors — mirror fixtures/affect_vad_derivation.json
// ---------------------------------------------------------------------------

const VECTORS: readonly AffectVadVector[] = [
  {
    id:          'default',
    description: 'Default AffectState — curiosity=0.5, engagement=0.5, uncertainty=0.2, rapport=0, energy=0.5',
    input:    { curiosity: 0.5, engagement: 0.5, uncertainty: 0.2, rapport: 0.0, energy: 0.5 },
    expected: { valence:   0.43333333, arousal: 0.425, dominance: 0.65 },
  },
  {
    id:          'all_max',
    description: 'All positive dimensions max, uncertainty zero.',
    input:    { curiosity: 1.0, engagement: 1.0, uncertainty: 0.0, rapport: 1.0, energy: 1.0 },
    expected: { valence:   1.0,        arousal: 0.75,  dominance: 1.0  },
  },
  {
    id:          'all_min_uncertain',
    description: 'All positive dimensions zero, uncertainty maxed.',
    input:    { curiosity: 0.0, engagement: 0.0, uncertainty: 1.0, rapport: 0.0, energy: 0.0 },
    expected: { valence:   0.0,        arousal: 0.25,  dominance: 0.0  },
  },
  {
    id:          'engagement_warm',
    description: 'Engaged + warm rapport + low uncertainty.',
    input:    { curiosity: 0.6, engagement: 0.9, uncertainty: 0.1, rapport: 0.8, energy: 0.7 },
    expected: { valence:   0.86666667, arousal: 0.525, dominance: 0.9  },
  },
  {
    id:          'stressed',
    description: 'High uncertainty, low rapport, low energy — overwhelmed state.',
    input:    { curiosity: 0.3, engagement: 0.2, uncertainty: 0.8, rapport: 0.0, energy: 0.2 },
    expected: { valence:   0.13333333, arousal: 0.375, dominance: 0.2  },
  },
  {
    id:          'energetic',
    description: 'High curiosity + high energy — alert / exploratory.',
    input:    { curiosity: 0.9, engagement: 0.6, uncertainty: 0.3, rapport: 0.4, energy: 0.9 },
    expected: { valence:   0.56666667, arousal: 0.75,  dominance: 0.65 },
  },
];

const EPSILON = 1e-5;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function makeState(input: AffectInput): AffectState {
  const s = new AffectState();
  s.curiosity   = input.curiosity;
  s.engagement  = input.engagement;
  s.uncertainty = input.uncertainty;
  s.rapport     = input.rapport;
  s.energy      = input.energy;
  return s;
}

function checkClose(field: string, got: number, want: number, epsilon: number): void {
  assert.ok(
    Math.abs(got - want) <= epsilon,
    `${field}: got ${got}, want ${want} (diff ${Math.abs(got - want)} > ${epsilon})`,
  );
}

// ---------------------------------------------------------------------------
// Tests — fixture-aligned derivation vectors
// ---------------------------------------------------------------------------

describe('affectStateToVad derivation vectors', () => {
  for (const v of VECTORS) {
    it(`${v.id} — ${v.description}`, () => {
      const state = makeState(v.input);
      const vad   = affectStateToVad(state);

      checkClose('valence',   vad.valence,   v.expected.valence,   EPSILON);
      checkClose('arousal',   vad.arousal,   v.expected.arousal,   EPSILON);
      checkClose('dominance', vad.dominance, v.expected.dominance, EPSILON);
    });
  }
});

// ---------------------------------------------------------------------------
// Output range — every component MUST be in [0, 1] for every fixture vector
// ---------------------------------------------------------------------------

describe('affectStateToVad output range', () => {
  for (const v of VECTORS) {
    it(`${v.id}: all components in [0, 1]`, () => {
      const vad = affectStateToVad(makeState(v.input));
      assert.ok(vad.valence   >= 0 && vad.valence   <= 1, `valence out of range: ${vad.valence}`);
      assert.ok(vad.arousal   >= 0 && vad.arousal   <= 1, `arousal out of range: ${vad.arousal}`);
      assert.ok(vad.dominance >= 0 && vad.dominance <= 1, `dominance out of range: ${vad.dominance}`);
    });
  }
});

// ---------------------------------------------------------------------------
// Purity — must not mutate the source AffectState
// ---------------------------------------------------------------------------

describe('affectStateToVad purity', () => {
  it('does not mutate the source AffectState', () => {
    const state = makeState({
      curiosity:   0.6,
      engagement:  0.9,
      uncertainty: 0.1,
      rapport:     0.8,
      energy:      0.7,
    });
    const before = {
      curiosity:   state.curiosity,
      engagement:  state.engagement,
      uncertainty: state.uncertainty,
      rapport:     state.rapport,
      energy:      state.energy,
    };
    affectStateToVad(state);
    assert.equal(state.curiosity,   before.curiosity);
    assert.equal(state.engagement,  before.engagement);
    assert.equal(state.uncertainty, before.uncertainty);
    assert.equal(state.rapport,     before.rapport);
    assert.equal(state.energy,      before.energy);
  });

  it('returns a fresh AffectVad on each call', () => {
    const state = new AffectState();
    const a: AffectVad = affectStateToVad(state);
    const b: AffectVad = affectStateToVad(state);
    assert.ok(a !== b, 'affectStateToVad must return a fresh object');
    assert.equal(a.valence,   b.valence);
    assert.equal(a.arousal,   b.arousal);
    assert.equal(a.dominance, b.dominance);
  });
});

File: go/tests/affect_vad_test.go
// affect_vad_test.go
//
// Validates AffectState.ToVad() — projection from the five engagement
// dimensions onto the Valence / Arousal / Dominance space.
//
// Derivation must be byte-identical across language ports; vectors live in
// fixtures/affect_vad_derivation.json. Float comparisons use epsilon = 1e-5
// (float32 precision; matches the fixture's declared epsilon).

package circleai_test

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

// ---------------------------------------------------------------------------
// Fixture types
// ---------------------------------------------------------------------------

type affectVadFixture struct {
	Epsilon float64         `json:"epsilon"`
	Vectors []affectVadCase `json:"vectors"`
}

type affectVadCase struct {
	ID          string     `json:"id"`
	Description string     `json:"description"`
	Input       affectDims `json:"input"`
	Expected    vadDims    `json:"expected"`
}

type vadDims struct {
	Valence   float64 `json:"valence"`
	Arousal   float64 `json:"arousal"`
	Dominance float64 `json:"dominance"`
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func loadAffectVadFixture(t *testing.T) affectVadFixture {
	t.Helper()
	path := filepath.Join(fixturesDir(t), "affect_vad_derivation.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read affect_vad_derivation.json: %v", err)
	}
	var fix affectVadFixture
	if err := json.Unmarshal(data, &fix); err != nil {
		t.Fatalf("failed to parse affect_vad_derivation.json: %v", err)
	}
	return fix
}

func assertVadDim(t *testing.T, id, name string, got float32, want float64, epsilon float64) {
	t.Helper()
	if math.Abs(float64(got)-want) > epsilon {
		t.Errorf("[%s] %s: got %v, want %v (epsilon %v)", id, name, got, want, epsilon)
	}
}

// ---------------------------------------------------------------------------
// Table-driven derivation tests
// ---------------------------------------------------------------------------

func TestAffectStateToVad_Vectors(t *testing.T) {
	const epsilon = 1e-5 // float32 precision; matches fixture epsilon

	fix := loadAffectVadFixture(t)

	if len(fix.Vectors) == 0 {
		t.Fatal("no vectors found in affect_vad_derivation.json")
	}

	for _, v := range fix.Vectors {
		v := v // capture
		t.Run(v.ID, func(t *testing.T) {
			state := circleai.AffectState{
				UserID:         "test",
				LastUpdatedUTC: time.Now().UTC(),
				Curiosity:      float32(v.Input.Curiosity),
				Engagement:     float32(v.Input.Engagement),
				Uncertainty:    float32(v.Input.Uncertainty),
				Rapport:        float32(v.Input.Rapport),
				Energy:         float32(v.Input.Energy),
			}

			vad := state.ToVad()

			assertVadDim(t, v.ID, "valence", vad.Valence, v.Expected.Valence, epsilon)
			assertVadDim(t, v.ID, "arousal", vad.Arousal, v.Expected.Arousal, epsilon)
			assertVadDim(t, v.ID, "dominance", vad.Dominance, v.Expected.Dominance, epsilon)
		})
	}
}

// ---------------------------------------------------------------------------
// Output-clamp guarantee — derived axes are bounded to [0.0, 1.0]
// ---------------------------------------------------------------------------

func TestAffectStateToVad_OutputClampedTo01(t *testing.T) {
	// Construct a state with values outside the documented [0,1] input range
	// so the unclamped formulas would overshoot. ToVad must still produce
	// outputs inside [0,1].
	state := circleai.AffectState{
		UserID:         "test",
		LastUpdatedUTC: time.Now().UTC(),
		Curiosity:      2.0,  // out of range — drives arousal above 1
		Engagement:     2.0,  // out of range — drives valence + dominance above 1
		Uncertainty:    -1.0, // out of range — drives all three above 1
		Rapport:        2.0,
		Energy:         2.0,
	}

	vad := state.ToVad()

	if vad.Valence < 0 || vad.Valence > 1 {
		t.Errorf("Valence not clamped to [0,1]: got %v", vad.Valence)
	}
	if vad.Arousal < 0 || vad.Arousal > 1 {
		t.Errorf("Arousal not clamped to [0,1]: got %v", vad.Arousal)
	}
	if vad.Dominance < 0 || vad.Dominance > 1 {
		t.Errorf("Dominance not clamped to [0,1]: got %v", vad.Dominance)
	}
}

func TestAffectStateToVad_OutputClampedToZero(t *testing.T) {
	// Mirror case — drive all formulas negative.
	state := circleai.AffectState{
		UserID:         "test",
		LastUpdatedUTC: time.Now().UTC(),
		Curiosity:      -2.0,
		Engagement:     -2.0,
		Uncertainty:    2.0, // (1 - 2.0) = -1.0 in valence + dominance
		Rapport:        -2.0,
		Energy:         -2.0,
	}

	vad := state.ToVad()

	if vad.Valence != 0 {
		t.Errorf("Valence: got %v, want 0 (lower clamp)", vad.Valence)
	}
	if vad.Arousal != 0 {
		t.Errorf("Arousal: got %v, want 0 (lower clamp)", vad.Arousal)
	}
	if vad.Dominance != 0 {
		t.Errorf("Dominance: got %v, want 0 (lower clamp)", vad.Dominance)
	}
}

// ---------------------------------------------------------------------------
// Default-state sanity check — uses NewAffectState constructor
// ---------------------------------------------------------------------------

func TestAffectStateToVad_DefaultStateMatchesFixture(t *testing.T) {
	const epsilon = 1e-5

	state := circleai.NewAffectState("default-user")
	vad := state.ToVad()

	// NewAffectState produces curiosity=0.5, engagement=0.5, uncertainty=0.2,
	// rapport=0, energy=0.5 — matches the "default_state" fixture vector.
	assertVadDim(t, "default_state", "valence", vad.Valence, 0.43333333, epsilon)
	assertVadDim(t, "default_state", "arousal", vad.Arousal, 0.425, epsilon)
	assertVadDim(t, "default_state", "dominance", vad.Dominance, 0.65, epsilon)
}
